lib/widgets/actions.dart

```diff
@@ -16,101 +16,171 @@ import '../model/narrow.dart';
 import 'dialog.dart';
 import 'store.dart';
 
-Future<void> markNarrowAsRead(
-  BuildContext context,
-  Narrow narrow,
-  bool useLegacy, // TODO(server-6)
-) async {
-  final store = PerAccountStoreWidget.of(context);
-  final connection = store.connection;
-  if (useLegacy) {
-    return await _legacyMarkNarrowAsRead(context, narrow);
-  }
-
-  // Compare web's `mark_all_as_read` in web/src/unread_ops.js
-  // and zulip-mobile's `markAsUnreadFromMessage` in src/action-sheets/index.js .
-  final zulipLocalizations = ZulipLocalizations.of(context);
-  final scaffoldMessenger = ScaffoldMessenger.of(context);
-  // Use [AnchorCode.oldest], because [AnchorCode.firstUnread]
-  // will be the oldest non-muted unread message, which would
-  // result in muted unreads older than the first unread not
-  // being processed.
-  Anchor anchor = AnchorCode.oldest;
-  int responseCount = 0;
-  int updatedCount = 0;
-
-  // Include `is:unread` in the narrow.  That has a database index, so
-  // this can be an important optimization in narrows with a lot of history.
-  // The server applies the same optimization within the (deprecated)
-  // specialized endpoints for marking messages as read; see
-  // `do_mark_stream_messages_as_read` in `zulip:zerver/actions/message_flags.py`.
-  final apiNarrow = narrow.apiEncode()..add(ApiNarrowIsUnread());
+Future<void> markNarrowAsRead(BuildContext context, Narrow narrow) async {
+  try {
+    final store = PerAccountStoreWidget.of(context);
+    final connection = store.connection;
+    final useLegacy = connection.zulipFeatureLevel! < 155; // TODO(server-6)
+    if (useLegacy) {
+      await _legacyMarkNarrowAsRead(context, narrow);
+      return;
+    }
 
-  while (true) {
-    final result = await updateMessageFlagsForNarrow(connection,
-      anchor: anchor,
+    // Compare web's `mark_all_as_read` in web/src/unread_ops.js
+    // and zulip-mobile's `markAsUnreadFromMessage` in src/action-sheets/index.js .
```

Comment:
This comment belongs in updateMessageFlagsStartingFromAnchor, rather than one of its callers; it's really about the logic that now lives in that function.